Complete restructuring of the SupplyChain module

aleeciu opened the pull request:
This pull request features a complete restructuring of the supply chain module, including:
- two subtle bug fixes
- the use of pymrio (https://pymrio.readthedocs.io/en/latest/intro.html) to download and handle multi-regional input output tables
- restructuring of operations which sped up the code substantially

aleeciu commented:
Hi @emanuel-schmid, in this pull request I am using the pymrio module (https://pymrio.readthedocs.io/en/latest/index.html) which is currently not included in the CLIMADA environment. How shall I add it?

emanuel-schmid commented:
Yes, this problem is currently not really solved. 

There are workarounds though: one can add a line `pip install pymrio` in `script/branches/Jenkinsfile`. With this, the environment is updated before running unit-tests on Jenkins. (I just did this and the tests ran through)

However, this file should actually remain unchanged in the develop branch and `requirements/env_climada.yaml` should be changed instead. You can update env_climada.yaml anytime in this branch but it won't have an effect on the environment, which is defined by the develop branch alone. On the other hand, the changes in the Jenkinsfile be undone before or after merging. (They don't hurt but the slow down the unittests for a couple of seconds for each commit.)

For now I suggest to proceed like this and I'll clean up the Jenkinsfile after the PR has been merged.

aleeciu commented:
Great, thanks @emanuel-schmid for the clarification. I will then update the `requirements/env_climada.yaml` in develop after merging. One more question: I keep getting a `readthedocs error` but I am not quite sure how to fix/interpret it.

aleeciu commented:
> > I keep getting a readthedocs error but I am not quite sure how to fix/interpret it.
> 
> @aleeciu Try merging the latest version of `develop` back into this branch

great, it worked thanks

aleeciu commented:
Jenkins has been running checks for almost six hours... any clue @emanuel-schmid @peanutfun ?

peanutfun commented:
The latest check failed: https://ied-wcr-jenkins.ethz.ch/job/petals_branches/job/PR-66/14/ I don't know why this is not reported back to GitHub 🤔 But the check is definitely not in progress anymore 😅

aleeciu commented:
> The latest check failed: https://ied-wcr-jenkins.ethz.ch/job/petals_branches/job/PR-66/14/ I don't know why this is not reported back to GitHub 🤔 But the check is definitely not in progress anymore 😅

ok cool, thanks. So it seems that the following test:
`climada_petals.engine.test.test_supplychain.TestSupplyChain.test_indirect_production_impact`
is failing, right? 

But it does not when I run it on my machine, and I really don't understand why. Does Jenkins gives fail if the test throws a `warning`?

emanuel-schmid commented:
No, it doesn't. But it [tells](https://ied-wcr-jenkins.ethz.ch/job/petals_branches/view/change-requests/job/PR-66/14/testReport/junit/climada_petals.engine.test.test_supplychain/TestSupplyChain/test_indirect_production_impact/) you what the problem is:

```
Traceback (most recent call last):
  File "/var/lib/jenkins/jobs/petals_branches/branches/PR-66/workspace/climada_petals/engine/test/test_supplychain.py", line 175, in test_indirect_production_impact
    self.assertAlmostEqual(sup.indir_prod_impt_mat.values.sum(), 7093283110973.164, places=3)
AssertionError: 7093283110973.162 != 7093283110973.164 within 3 places (0.001953125 difference)
```

These rounding differences are quite common and the reason could be anything from dependency library over compiler to processor, afaik. On the bright side: easy to fix 😁

aleeciu commented:
thanks for your hints - I fixed it now - although I am pretty sure @peanutfun will request extensive changes to these tests :-)

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+        recix = 0
+    else:
+        with warnings.catch_warnings():
+            # catch the divide by zero warning
+            # we deal wit that by setting to 0 afterwards
+            warnings.simplefilter("ignore")
```
Only ignore the divide by zero warning
```suggestion
            # Ignore devide by zero warning, we set to 0 afterwards
            warnings.filterwarnings("ignore", message="divide by zero")
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+            # we deal wit that by setting to 0 afterwards
+            warnings.simplefilter("ignore")
+            recix = 1 / x
+        recix[recix == np.inf] = 0
```
Test for negative inf as well?
```suggestion
        recix[recix.isinf()] = 0
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+            warnings.simplefilter("ignore")
+            recix = 1 / x
+        recix[recix == np.inf] = 0
+        recix = recix.reshape((-1, 1))
```
Seems unnecessary given the broadcasting rules. Tests pass without this line.
```suggestion
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+        recix = recix.reshape((-1, 1))
+
+    if isinstance(Z, pd.DataFrame):
+        return pd.DataFrame(Z.values * recix, index=Z.index, columns=Z.columns)
```
```suggestion
        return pd.DataFrame(Z.to_numpy() * recix, index=Z.index, columns=Z.columns)
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+    Returns
+    -------
+    pandas.DataFrame or numpy.array
+        Industry output x as column vector
+        The type is determined by the type of G. If DataFrame index as G
```
```suggestion
        Industry output x as column vector.
        Same type as input parameter ``G``.
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+        mriot.x = pymrio.calc_x(mriot.Z, mriot.Y)
 
-        """
+    return mriot
```
Variable possibly unbound here
```suggestion
    else:
        raise RuntimeError(f"Unknown mriot_type: {mriot_type}")

    return mriot
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+                * self.conv_fac()
+            )
+
+        self.indir_prod_impt_eai = self.indir_prod_impt_mat.T.dot(impact.frequency)
```
Same as above, why not return the value?

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+
+        self.indir_prod_impt_eai = self.indir_prod_impt_mat.T.dot(impact.frequency)
+
+    def calc_total_production_impacts(self, impact):
+        """Calculate total production impacts."""
+        self.tot_prod_impt_mat = self.dir_prod_impt_mat.add(self.indir_prod_impt_mat)
+        self.tot_prod_impt_eai = self.tot_prod_impt_mat.T.dot(impact.frequency)
```
Same as above, why not expect `<in>dir_prod_impt_mat` as arguments? This method never checks if the attributes are really set.

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+        self,
+        impact,
+        exposure,
+        impacted_secs=None,
+        io_approach=None,
+        stock_to_prod_shock=None,
```
Please add documentation for these parameters

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+        self.calc_total_production_impacts(impact)
+
+    def calc_matrixes(self, io_approach):
+        """
+        Build technical coefficient and Leontief inverse matrixes (if Leontief approach)
+        or allocation coefficients and Ghosh matrixes (if Ghosh approach).
+        """
```
Please add documentation for `io_approach` parameter

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+        self.coeffs = coeff_func(self.mriot.Z, self.mriot.x)
+        self.inverse = inv_func(self.coeffs)
+
+    def conv_fac(self):
```
Unnecessary abbreviation?
```suggestion
    def conversion_factor(self):
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+                )
+                * self.conv_fac()
+            )
+
```
Need to handle unknown approach?

```suggestion
        else:
            raise RuntimeError(f"Unknown io_approach: {io_approach}")

```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,479 @@
+        for exp_regid in exposure.gdf.region_id.unique():
+            exp_bool = exposure.gdf.region_id == exp_regid
+            tot_value_reg_id = exposure.gdf[exp_bool].value.sum()
+            # consider using impact.impact_reg_agg when merged
+            tot_imp_reg_id = impact.imp_mat[:, np.where(exp_bool)[0]].sum(1)
```
`Impact.impact_at_reg` is available now!

peanutfun reviewed (changes requested):
I found a lot of small issues, but overall, this is looking good! 🙌  I have a few general comments:

* The tutorial seems to work well and gives a nice overview. However, most computed values seem to be zero. It would be great if you could adapt the tutorial to focus on a particularly impacted sector, in order for readers to literally *see something* (most displayed data frames only show zeros).
* I do not understand the tests as they rely on real data. Can you add tests where you use self-made, dummy data for testing the calculations?
* The `SupplyChain` class mostly stores computed data in attributes and uses these attributes in the logically following calculation. But many of these calculations never check if the attribute actually exists. This will lead to weird errors if the methods are called in the wrong order. I would prefer returning the values to the user and then having the user insert them in the next function. This ensures that the data required for the calculation is always there.
* Many methods and functions are named `calc_XYZ`. I personally would prefer removing this prefix. It is usually clear that the functions calculate something. This way, the names get shorter and less abbreviations need to be used.
* Several functions try to work with both numpy and pandas data structures. It feels to me like this may introduce subtle bugs that are difficult to test and debug. Is this at all necessary?

peanutfun commented:
@aleeciu How do we proceed here?

peanutfun commented:
> Also, there is a fork of this fork https://github.com/CLIMADA-project/climada_petals/tree/feature/climada_boario and I am torned whether adding those changes to this PR or not.

@aleeciu The smaller the changes, the better. I see no conflicts between the `supplychain` and `climada_boario` branches.  I suggest merging this PR first. After that, reviewing `climada_boario` separately should be much easier.

aleeciu commented on climada_petals/engine/supplychain.py:
```diff
@@ -22,223 +22,446 @@
+            Technical (if Leontief, A) or allocations (if Ghosh, B) coefficients matrix
+    inverse : pd.DataFrame
+            Leontief (L) or Ghosh (G) inverse matrix
+    dir_prod_impt_mat : pd.DataFrame
+            Direct production impact for each country, sector and event
+    dir_prod_impt_eai : pd.DataFrame
+            Expected direct production impact for each country and sector
+    indir_prod_impt_mat : pd.DataFrame
+            Indirect production impact for each country, sector and event
+    indir_prod_impt_mat : pd.DataFrame
+            Expected indirect production impact for each country and sector
+    tot_prod_impt_mat : pd.DataFrame
+            Total production impact for each country, sector and event
+    tot_prod_impt_eai : pd.DataFrame
+            Expected total production impact for each country and sector
```
After a bilateral discussion with @peanutfun , we decided not to implement this since indeed some - but not all - DataFrames have the same structure

aleeciu commented:
I updated the branch based on the inputs from @peanutfun (and also applied some other changes but nothing major):

- better tutorial, no zeros everywhere and I hope more detailed explanation
- better tests, when I could I built dummy tables and checked results against expected ones. In some cases - however - operations are not straightforward and I would not know what to do rather than rewriting code to have these "expected" results
- rewrote the docstring and cleaned up all the unused / undeclared attributed
- I am getting some test errors on Jenkins (but not locally) - but if I understand they are not related to the module develop in this pull-request (https://ied-wcr-jenkins.ethz.ch/job/petals_branches/view/change-requests/job/PR-66/lastCompletedBuild/testReport/)

On my side - I'd inclined to say we are there, but of course happy to get more feedbacks.

peanutfun commented:
Thanks a lot @aleeciu! I'll have another thorough look next week

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+
+    if isinstance(x, (pd.DataFrame, pd.Series)):
+        x = x.to_numpy()
+    if (type(x) is not np.ndarray) and (x == 0):
```
```suggestion
    if not isinstance(x, np.ndarray) and (x == 0):
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+    col_sectors : int
+        Column's position of sectors' names
+    rows_data : (int, int)
+        Tuple of integers with positions of rows 
```
```suggestion
        Tuple of integers with positions of rows
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+        Tuple of integers with positions of rows 
+        containing the MRIOT data
+    cols_data : (int, int)
+        Tuple of integers with positions of columns 
```
```suggestion
        Tuple of integers with positions of columns
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+    if mriot_type == "EXIOBASE3":
+        return f"IOT_{mriot_year}_ixi.zip"
+
+    elif mriot_type == "WIOD16":
```
```suggestion
    if mriot_type == "WIOD16":
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+    elif mriot_type == "OECD21":
+        return f"ICIO2021_{mriot_year}.csv"
+
+    else:
+        raise ValueError("Unknown MRIOT type")
```
```suggestion
    raise ValueError("Unknown MRIOT type")
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
-class SupplyChain():
+
+def download_mriot(mriot_type, mriot_year, download_dir):
+    """Download EXIOBASE3, WIOD16 or OECD21 Multi-Regional Input Output Tables 
```
```suggestion
    """Download EXIOBASE3, WIOD16 or OECD21 Multi-Regional Input Output Tables
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+
+def download_mriot(mriot_type, mriot_year, download_dir):
+    """Download EXIOBASE3, WIOD16 or OECD21 Multi-Regional Input Output Tables 
+    for specific years. 
```
```suggestion
    for specific years.
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+                mriot.x : industry or total output
+                mriot.meta : metadata
+    secs_exp : pd.DataFrame
+            Exposure dataframe of each country/sector in the MRIOT. Columns are the 
```
```suggestion
            Exposure dataframe of each country/sector in the MRIOT. Columns are the
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+            Exposure dataframe of each country/sector in the MRIOT. Columns are the 
+            same as the chosen MRIOT.
+    secs_imp : pd.DataFrame
+            Impact dataframe for the directly affected countries/sectors for each event with impacts.
+            Columns are the same as the chosen MRIOT and rows are the hazard events ids.
```
```suggestion
            Impact dataframe for the directly affected countries/sectors for each event with
            impacts. Columns are the same as the chosen MRIOT and rows are the hazard events ids.
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+            hazard events ids.
+    inverse : dict
+            Dictionary with keys being the chosen approach (ghosh, leontief or eeioa)
+            and values the Leontief (L, if approach is leontief or eeioa) or Ghosh (G, if 
```
```suggestion
            and values the Leontief (L, if approach is leontief or eeioa) or Ghosh (G, if
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+    def from_mriot(
+        cls, mriot_type, mriot_year, mriot_dir=MRIOT_DIRECTORY, del_downloads=True
+    ):
+        """Download, parse and read WIOD16, EXIOBASE3, or OECD21 Multi-Regional 
```
```suggestion
        """Download, parse and read WIOD16, EXIOBASE3, or OECD21 Multi-Regional
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+
+        return cls(mriot=mriot)
+
+    def calc_shock_to_sectors(self, 
+                              exposure, 
+                              impact, 
+                              impacted_secs=None, 
```
```suggestion
    def calc_shock_to_sectors(self,
                              exposure,
                              impact,
                              impacted_secs=None,
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+                              shock_factor=None
+                              ):
+        """Calculate exposure, impact and shock at the sectorial level.
+        This function translate spatially-distrubted exposure and impact 
```
```suggestion
        This function translate spatially-distrubted exposure and impact
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+                              ):
+        """Calculate exposure, impact and shock at the sectorial level.
+        This function translate spatially-distrubted exposure and impact 
+        information into exposure and impact of MRIOT's country/sectors and 
```
```suggestion
        information into exposure and impact of MRIOT's country/sectors and
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+            self.secs_shock[self.secs_shock > 1] = 1
+
+    def calc_matrices(self, io_approach):
+        """Build technical coefficient and Leontief inverse matrixes 
```
```suggestion
        """Build technical coefficient and Leontief inverse matrixes
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+
+    def calc_matrices(self, io_approach):
+        """Build technical coefficient and Leontief inverse matrixes 
+        (if leontief or eeioa approach) or allocation coefficients and 
```
```suggestion
        (if leontief or eeioa approach) or allocation coefficients and
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -19,226 +19,580 @@
+                    exposure=None,
+                    impact=None,
+                    impacted_secs=None):
+        """Calculate indirect production impacts based on to the 
```
```suggestion
        """Calculate indirect production impacts based on to the
```

peanutfun commented on climada_petals/engine/supplychain.py:
```diff
@@ -250,101 +604,72 @@ def calc_indirect_impact(self, io_approach='ghosh'):
+        if not idx_country.size > 0.0:
+            mriot_reg_name = "ROW"
+
+        return mriot_reg_name
```
```suggestion
        return mriot_reg_name

```

aleeciu commented:
Many thanks @peanutfun; I am confused because I saw some suggested changes from your side but now they are marked as resolved. It's all good for me if you implemented them directly, just please let me know if changes are still expected from my side.

ps
I did not mean to urge you few days ago by reminding you about the review - I just pressed the button by mistake :-)

peanutfun commented:
@aleeciu I now marked the comments as resolved. Some were outdated and for others I am not familiar enough with the module to mess with it now. If you are fine with how it works, I think we can merge 🙂

aleeciu commented:
we can merge from my side, and we still have #81 which will bring changes to this module
